### PR TITLE
feat: add parallel runner

### PR DESCRIPTION
**Pull request by KveinAxel**

### CODE REVIEW

File: sqllogictest/src/runner.rs [hunk 1]
```diff
@@ -33,6 +34,19 @@ pub trait AsyncDB: Send {
...
+    async fn run_task<F>(futures: Vec<F>)
+    where
+        F: Future + Send + 'static,
+        <F as Future>::Output: Send + 'static,
+    {
+        let handles = futures.into_iter().map(|future| tokio::spawn(future));
+
+        for handle in handles {
+            handle.await.unwrap();
+        }
```
Comment (wangrunji0408) [inline sqllogictest/src/runner.rs:48]: It seems that `tokio::spawn` is not necessary here. Can be replaced by `join_all` (?

File: sqllogictest/src/runner.rs [hunk 2]
```diff
@@ -328,6 +342,135 @@ impl<D: AsyncDB> Runner<D> {
...
+        ParallelRunner { db, hosts }
+    }
+
+    /// parse slt files and create DB connection for multiple target. Here we use the conn_builder
+    /// parameter function for connection creating, the function accept two parameter which is host
+    /// and dbname in order.
+    pub async fn prepare_async<F, Fut>(
+        &mut self,
+        glob: &str,
+        conn_builder: F,
```
Comment (wangrunji0408) [inline sqllogictest/src/runner.rs:369]: move to `AsyncDB` as an async constructor fn?

File: sqllogictest/src/runner.rs [hunk 3]
```diff
@@ -328,6 +342,135 @@ impl<D: AsyncDB> Runner<D> {
...
+    }
+
+    /// aceept the tasks, spawn jobs task to run slt test. the tasks are (AsyncDB, slt filename)
+    /// pairs.
+    pub async fn run_parallel_async(
+        &self,
+        mut tasks: Vec<(D, String)>,
+        jobs: usize,
+    ) -> Result<(), TestError> {
+        let chunk_size = (tasks.len() as f64 / jobs as f64).ceil() as usize;
```
Comment (wangrunji0408) [inline sqllogictest/src/runner.rs:425]: prefer integer operations to floating point
```suggestion
        let chunk_size = (tasks.len() + jobs - 1) / jobs;
```

File: sqllogictest/src/runner.rs [hunk 4]
```diff
@@ -328,6 +329,130 @@ impl<D: AsyncDB> Runner<D> {
...
+        let mut chunked_tasks = vec![];
+        loop {
+            if tasks.len() > chunk_size {
+                chunked_tasks.push(tasks.drain((tasks.len() - chunk_size)..).collect_vec());
+            } else {
+                chunked_tasks.push(tasks.drain(..).collect_vec());
+                break;
+            }
+        }
```
Comment (wangrunji0408) [inline sqllogictest/src/runner.rs:434]: There is a function [buffer_unordered(n)](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.buffer_unordered) that can dynamically allocate tasks into n parallel jobs.
I think it's better than static allocation.

File: sqllogictest/src/runner.rs [hunk 5]
```diff
@@ -328,6 +329,130 @@ impl<D: AsyncDB> Runner<D> {
     }
 }
 
+/// A parallel Sqllogictest runner
+///
+/// before beginning the runner, we should use prepare(_async) to parse slt files and split task and
+/// then we can pass the tasks to runner for parallel running.
```
Comment (wangrunji0408) [inline sqllogictest/src/runner.rs:335]: Is it possible to merge the prepare and run function into one?
One step further, is it possible not to introduce a new runner?

File: sqllogictest/Cargo.toml [hunk 6]
```diff
@@ -20,3 +20,4 @@ itertools = "0.10"
 log = "0.4"
 tempfile = "3"
 thiserror = "1"
+futures = "0.3"
```
Comment (skyzh) [inline sqllogictest/Cargo.toml:23]: If we are eventually use futures, we should remove futures_lite.

File: sqllogictest/src/runner.rs [hunk 7]
```diff
@@ -303,12 +306,78 @@ impl<D: AsyncDB> Runner<D> {
...
+                let db = conn_builder(target, db_name).await;
+                let mut tester = Runner::new(db);
+                let filename = file.to_string_lossy().to_string();
+                tester.run_file_async(filename).await
+            })
+        }
+
+        let mut tasks = stream::iter(tasks).buffer_unordered(jobs);
+        while let Some(result) = tasks.next().await {
+            result?;
```
Comment (skyzh) [inline sqllogictest/src/runner.rs:363]: The error returned here might be confusing to developers. As we are run-in a lot of test cases in parallel, we should let developer know which test case failed exactly. Better to have a new error type with which test case failed.

**Review by skyzh — APPROVED** [review]
Rest LGTM, please also resolve comments from other developers.

**Review by wangrunji0408 — APPROVED** [review]
LGTM!

**Pull request merged**